# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/wasm/WasmInstance.cpp
# Commit: 546d74acfb81
# Full Hash: 546d74acfb81497dc00809fd4e4096edfe857ef5
# Author: Ryan Hunt <rhunt@eqrion.net>
# Date: 2020-08-19 10:01:16
# Regressor Bug: 1561521
# File Overlap Count: 2
# Description:
#   Bug 1561521 - Add non-nullable references. r=lth
#   
#   This commit adds decoding, validation, and JS-API type-checking for
#   non-nullable references.
#   
# ==============================================================================

diff -r e055945526a2 -r 546d74acfb81 js/src/wasm/WasmInstance.cpp
--- a/js/src/wasm/WasmInstance.cpp	Tue Aug 18 16:58:55 2020 +0000
+++ b/js/src/wasm/WasmInstance.cpp	Tue Aug 18 16:58:59 2020 +0000
@@ -200,6 +200,15 @@
     case ValType::V128:
       MOZ_CRASH("unexpected v128 in ToWebAssemblyValue");
     case ValType::Ref:
+#ifdef ENABLE_WASM_GC
+      if (!type.isNullable() && val.isNull()) {
+        JS_ReportErrorNumberUTF8(cx, GetErrorMessage, nullptr,
+                                 JSMSG_WASM_BAD_REF_NONNULLABLE_VALUE);
+        return false;
+      }
+#else
+      MOZ_ASSERT(type.isNullable());
+#endif
       switch (type.refTypeKind()) {
         case RefType::Func:
           return ToWebAssemblyValue_funcref<Debug>(cx, val, (void**)loc);